crates/hir_ty/src/diagnostics/match_check/usefulness.rs

```diff
@@ -331,8 +331,7 @@ impl<'a> MatchCheckCtx<'a> {
 
     #[track_caller]
     pub(super) fn bug(&self, info: &str) -> ! {
```

Comment:
(note: I haven't checked where this is called, but in general we want to be a bit more resilient about panicking on errors than rustc: Showing some inaccurate information in the IDE is often better than breaking completely. Hence the `always!` macro etc. that panic on debug assertions, but only log and continue in release builds.)

Reply:
Interesting. This is worth some experimenting with fallible implementation then including it into future libraryfied match checking.
For now it just relies on correct lowering and filtering out erroneous input in `match_check::PatCtxt::lower_pattern`. But in case of bugs, yeah.. that's bad.

Reply:
Note, it's not worth making functions return results to avoid assertions. It's just, *if* there's a straightforward way to e.g. return no errors instead of panicking on an assertion, we should do that.